### PR TITLE
Go error tests

### PR DESCRIPTION
**Pull request by pongad**
@shinfan @garrettjonesgoogle Please take a look at MVVM and testing code. Most of it was me pulling things out of the Java implementation.

@jba Please take a look at the Go surface. The mock type is slightly renamed `mockLibraryServer` is the type and `mockLibrary` the variable. I'm currently using reflection to initialize the request object to zero value. This is obviously undesirable, but it should be solved quite naturally once the happy-case tests land (successfully receive replies from the mock services). We will have to initialize the request to non-zero state to properly test the happy-case. Once we do that, inits to zero state should be easy.

The test has also caught a bug in iteration code. Previously InternalFetch returns fields of response along with errors. In case of errors though, the response is a nil pointer and accessing its fields causes a panic. The bug has been fixed here.

@jmuk @bjwatson I'm not sure what the timeline for tests is for other languages (to the best of my knowledge, it's not yet complete). I'd appreciate you taking a glace at `MockServiceTransformer`, as we can use it to implement testing for other languages.

### CODE REVIEW

File: src/main/resources/com/google/api/codegen/go/mock.snip [hunk 1]
```diff
@@ -127,3 +174,58 @@
...
+        {@test.mockServiceVarName}.err = expErr
+
+        c, err := {@test.serviceConstructorName}(context.Background(), clientOpt)
+        if err != nil {
+            t.Fatal(err)
+        }
+
+        var req {@test.requestTypeName}
+
+        # TODO(pongad): Use struct initializer instead of reflection.
```
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/go/mock.snip]: Shouldn't this use the init code stuff?

File: src/main/java/com/google/api/codegen/viewmodel/testing/MockCombinedView.java [hunk 2]
```diff
@@ -30,7 +30,9 @@
 
   public abstract List<MockServiceImplView> serviceImpls();
 
-  // TODO(pongad): Add the tests
+  public abstract List<GapicSurfaceTestClassView> tests();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/viewmodel/testing/MockCombinedView.java]: I think `testClasses` (or `testFiles`) would be better, because `tests` is ambiguous with `testCases`

File: src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import com.google.api.codegen.viewmodel.ViewModel;
+import com.google.api.codegen.viewmodel.testing.MockServiceImplView.Builder;
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class GapicSurfaceTestFileView implements ViewModel {
+
+  public abstract FileHeaderView fileHeader();
+
+  public abstract GapicSurfaceTestClassView test();
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java]: I think `testClass` would be a better name because `test` is ambiguous with `testCase`

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java [hunk 4]
```diff
@@ -285,19 +280,28 @@ private GapicSurfaceTestClassView createUnitTestClassView(SurfaceTransformerCont
...
           testCaseViews.add(
-              createTestCaseView(
+              mockServiceTransformer.createTestCaseView(
                   methodContext,
                   testNameTable,
-                  flatteningGroup.getFlattenedFieldConfigs().values()));
+                  flatteningGroup.getFlattenedFieldConfigs().values(),
+                  methodConfig.isPageStreaming()
+                      ? ApiMethodType.PagedFlattenedMethod
+                      : ApiMethodType.FlattenedMethod,
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java]: Create a local var for the method type - the embedded structure is harder to follow. Also, FYI, I am renaming `ApiMethodType` to `ClientMethodType` in my PR https://github.com/googleapis/toolkit/pull/755 .

File: src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java [hunk 5]
```diff
@@ -73,4 +106,244 @@
...
+              .registerFunctionName(namer.getServerRegisterFunctionName(service))
+              .build();
+      mockServices.add(mockService);
+    }
+
+    return mockServices;
+  }
+
+  // TODO(shinfan): Convert to use TestMethodView.
+  public GapicSurfaceTestCaseView createTestCaseView(
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java]: Is Shin's TODO worth taking care of now?

File: src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java [hunk 6]
```diff
@@ -73,4 +106,244 @@
...
+        responseTypeName =
+            namer.getAndSavePagedResponseTypeName(
+                method, methodContext.getTypeTable(), resourcesField);
+        break;
+      default:
+        responseTypeName =
+            methodContext.getTypeTable().getAndSaveNicknameFor(method.getOutputType());
+    }
+
+    String surfaceMethodName;
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java]: call this `clientMethodName` (as part of the transition of "Api" to "Client")

File: src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java [hunk 7]
```diff
@@ -57,6 +69,7 @@ private MockCombinedView generateMockServiceView(
       Model model, ApiConfig apiConfig, SurfaceNamer namer) {
     ModelTypeTable typeTable = GoGapicSurfaceTransformer.createTypeTable();
     List<MockServiceImplView> impls = new ArrayList<>();
+    List<GapicSurfaceTestClassView> tests = new ArrayList<>();
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java]: s/tests/testClasses/

File: src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java (hunk 5 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java]: @pongad do you mind taking care of this now or I can fix this TODO after this PR is merged. I am good with either way.

File: src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java [hunk 8]
```diff
@@ -68,13 +81,50 @@ private MockCombinedView generateMockServiceView(
...
+  private List<GapicSurfaceTestCaseView> createTestCaseViews(SurfaceTransformerContext context) {
+    ArrayList<GapicSurfaceTestCaseView> testCaseViews = new ArrayList<>();
+    for (Method method : context.getSupportedMethods()) {
+      MethodConfig methodConfig = context.getMethodConfig(method);
+      testCaseViews.add(
+          mockServiceTransformer.createTestCaseView(
+              context.asRequestMethodContext(method),
+              new SymbolTable(),
+              Collections.<FieldConfig>emptyList(),
+              methodConfig.isPageStreaming()
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java]: Make this a local var

File: src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java [hunk 9]
```diff
@@ -68,13 +81,50 @@ private MockCombinedView generateMockServiceView(
...
+      SurfaceTransformerContext context =
+          SurfaceTransformerContext.create(
+              service,
+              apiConfig,
+              GoGapicSurfaceTransformer.createTypeTable(),
+              namer,
+              featureConfig);
+      tests.add(
+          GapicSurfaceTestClassView.newBuilder()
+              .apiSettingsClassName("")
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java]: This is okay but not very readable to pass empty strings here. If the field is not used by Go, we can mark it nullable in the view class or we can have some constants such as UNUSED

File: src/main/java/com/google/api/codegen/viewmodel/testing/MockCombinedView.java (hunk 2 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/viewmodel/testing/MockCombinedView.java]: testClasses it is

File: src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java (hunk 3 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java]: Done

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java (hunk 4 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTestTransformer.java]: Refactored back to the original.

File: src/main/resources/com/google/api/codegen/go/mock.snip (hunk 1 above)
Comment (pongad) [inline src/main/resources/com/google/api/codegen/go/mock.snip]: Init code for Go doesn't work yet. I'd prefer to do it in a separate PR. This PR is already quite large and touches some shared code, so merges are getting expensive.

File: src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java (hunk 5 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java]: It is definitely worth taking care of. As with Go init code, I'd prefer to do it in separate PRs.

File: src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java (hunk 6 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java]: Refactored out of existence.

File: src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java (hunk 9 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java]: Changed to string literal `"UNUSED"`

**Comment by pongad** [comment]
@garrettjonesgoogle @shinfan PTAL

File: src/test/java/com/google/api/codegen/testdata/go_mock_library.baseline [hunk 10]
```diff
@@ -277,13 +285,492 @@ type mockLabeler struct {
...
+        t.Fatal(err)
+    }
+
+    var req *librarypb.CreateShelfRequest
+
+    reflect.ValueOf(&req).Elem().Set(reflect.New(reflect.TypeOf(req).Elem()))
+
+    _, err = c.CreateShelf(context.Background(), req)
+
+    if desc := grpc.ErrorDesc(err); desc != expErr.Error() {
```
Comment (jba) [inline src/test/java/com/google/api/codegen/testdata/go_mock_library.baseline]: Compare codes instead of message strings.

Comment (pongad) [inline src/test/java/com/google/api/codegen/testdata/go_mock_library.baseline]: Comparing code is a good point. Is there a reason to note compare both?

Comment (jba) [inline src/test/java/com/google/api/codegen/testdata/go_mock_library.baseline]: The message should be just for humans. Code should be able to change the message without breaking tests.

File: src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java [hunk 11]
```diff
@@ -958,6 +958,10 @@ public String injectRandomStringGeneratorCode(String randomString) {
     return getNotImplementedString("SurfaceNamer.getRandomStringValue");
   }
 
+  /** Function used to register the GRPC server. */
+  public String getServerRegisterFunctionName(Interface service) {
+    return getNotImplementedString("SurfaceNamer.getServerRegisterFunctionName");
+  }
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java]: nit: empty line between methods

File: src/main/resources/com/google/api/codegen/go/mock.snip (hunk 1 above)
Comment (garrettjonesgoogle) [inline src/main/resources/com/google/api/codegen/go/mock.snip]: Ahh ok. Could you change the comment to be something like "Use the init code generator here"?

**Comment by shinfan** [comment]
LGTM after one small nit

File: src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+import com.google.api.tools.framework.model.TypeRef;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/** GapicTestTransformer contains helper methods useful for creating mock and test views. */
+public class GapicTestTransformer {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java]: None of our other transformers start with `Gapic`. I would suggest using `Client` in place of `Gapic` in general for class names.

With respect to this specific class, since you went and removed `createTestCaseView`, it seems like the original name of `MockServiceTransformer` makes a lot more sense.

File: src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+              .implName(namer.getMockGrpcServiceImplName(service))
+              .registerFunctionName(namer.getServerRegisterFunctionName(service))
+              .build();
+      mockServices.add(mockService);
+    }
+
+    return mockServices;
+  }
+
+  public InitCodeContext createRequestInitCodeContext(
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java]: I think this should actually live in `InitCodeTransformer`. There's nothing really test-specific about this.

File: src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+package com.google.api.codegen.viewmodel.testing;
+
+import com.google.api.codegen.SnippetSetRunner;
+import com.google.api.codegen.viewmodel.FileHeaderView;
+import com.google.api.codegen.viewmodel.ViewModel;
+import com.google.api.codegen.viewmodel.testing.MockServiceImplView.Builder;
+import com.google.auto.value.AutoValue;
+
+@AutoValue
+public abstract class GapicSurfaceTestFileView implements ViewModel {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java]: Don't prefix this with `Gapic` - I would suggest `ClientTestFileView`. I do see that there are some existing view model classes with this name, but I'd like to get them renamed too (I won't make you do it in this PR).

File: src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java (hunk 11 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java]: I thought the formatter would take care of this. How odd!

File: src/test/java/com/google/api/codegen/testdata/go_mock_library.baseline (hunk 10 above)
Comment (pongad) [inline src/test/java/com/google/api/codegen/testdata/go_mock_library.baseline]: Good point, Done

File: src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java [hunk 15]
```diff
@@ -37,8 +54,10 @@
 
   private final FileHeaderTransformer fileHeaderTransformer =
       new FileHeaderTransformer(new GoImportTransformer());
-  private final MockServiceTransformer mockServiceTransformer = new MockServiceTransformer();
+  private final GapicTestTransformer mockServiceTransformer =
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java]: Changed name of class back to MockServiceTransformer

File: src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java [hunk 16]
```diff
@@ -68,13 +88,100 @@ private MockCombinedView generateMockServiceView(
...
+              service,
+              apiConfig,
+              GoGapicSurfaceTransformer.createTypeTable(),
+              namer,
+              featureConfig);
+      testClasses.add(
+          GapicSurfaceTestClassView.newBuilder()
+              .apiSettingsClassName("UNUSED")
+              .apiClassName(namer.getApiWrapperClassName(service))
+              .name("UNUSED")
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/go/GoGapicSurfaceTestTransformer.java]: Done

File: src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java (hunk 12 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java]: Changed back to MockServiceTransformer

File: src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java (hunk 13 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/GapicTestTransformer.java]: Done

File: src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java (hunk 14 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/viewmodel/testing/GapicSurfaceTestFileView.java]: Made the change for all files. With the right automation, it's easy :)

File: src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java [hunk 17]
```diff
@@ -122,7 +122,7 @@ public GapicTestTransformer(ValueProducer valueProducer) {
     return mockServices;
   }
 
-  public InitCodeContext createRequestInitCodeContext(
+  private InitCodeContext createRequestInitCodeContext(
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/MockServiceTransformer.java]: If you put this in InitCodeTransformer, why didn't you remove it from here?

**Comment by garrettjonesgoogle** [comment]
LGTM except for one concern about a duplicated method - either address it or postpone fixing it for later, but I don't want to block this PR another day.

**Comment by pongad** [comment]
I have missed it earlier. Fixing.

**Comment by bjwatson** [comment]
FYI @jcanizales, who is handling automated tests for MTV right now.
